### PR TITLE
rss open source library x apollo 

### PR DESCRIPTION
**Pull request by l1212s**
RSS open source library is hosted in https://github.com/intel/ad-rss-lib
Here is the patch to intercept it with Apollo.  
With help of this patch, Apollo will fetch the RSS open source library from remote repository and build it automatically. 
The patch has been pass unit test and integrated test.

### CODE REVIEW

File: modules/planning/tasks/rss/decider_rss.cc [hunk 1]
```diff
@@ -228,22 +245,23 @@ Status RssDecider::Process(Frame *frame,
...
+      double(accelerationRestriction.lateralLeftRange.minimum));
   reference_line_info->mutable_rss_info()->set_acc_lat_left_range_maximum(
-      accelerationRestriction.lateralLeftRange.maximum);
+      double(accelerationRestriction.lateralLeftRange.maximum));
   reference_line_info->mutable_rss_info()->set_acc_lat_right_range_minimum(
-      accelerationRestriction.lateralRightRange.minimum);
+      double(accelerationRestriction.lateralRightRange.minimum));
   reference_line_info->mutable_rss_info()->set_acc_lat_right_range_maximum(
-      accelerationRestriction.lateralRightRange.maximum);
+      double(accelerationRestriction.lateralRightRange.maximum));
```
Comment (xiaoxq) [inline modules/planning/tasks/rss/decider_rss.cc]: 1. This file failed to pass lint check. You may try `apollo.sh lint` to find these errors, or use tool `scripts/clang-format.sh` to fix it automatically.
2. Use `static_cast<double>` to cast scala types.

File: third_party/rss_lib.BUILD [hunk 2]
```diff
@@ -0,0 +1,70 @@
+# ----------------- BEGIN LICENSE BLOCK ---------------------------------
+#
+# INTEL CONFIDENTIAL
```
Comment (ycool) [inline third_party/rss_lib.BUILD]: Can you help to update the license with Apache license?

File: modules/planning/tasks/rss/decider_rss.cc (hunk 1 above)
Comment (l1212s) [reply modules/planning/tasks/rss/decider_rss.cc]: @xiaoxq ,  thanks for your comment.  this issue is resolved in the patch "rss-open-source-library-fix-cpplint-issue"

File: third_party/rss_lib.BUILD (hunk 2 above)
Comment (l1212s) [reply third_party/rss_lib.BUILD]: @ycool thanks for your comments, this issue is resolved in the patch "remove-rss-lib-apollo-build-script-legal-disclaimer"

File: modules/planning/tasks/rss/decider_rss.cc [hunk 3]
```diff
@@ -125,94 +132,104 @@ Status RssDecider::Process(Frame *frame,
...
+         << double(laneSegment.width.minimum)
+         << " laneSegment.width.maximum = "
+         << double(laneSegment.width.maximum);
+
+  ad_rss::world::WorldModel worldModel;
+  worldModel.egoVehicle = ad_rss::objectAsEgo(followingObject);
   scene.object = leadingObject;
   scene.egoVehicleRoad = roadArea;
   worldModel.scenes.push_back(scene);
   worldModel.timeIndex = 1;
```
[the remark this answers is not in the document]
Comment (ycool) [reply modules/planning/tasks/rss/decider_rss.cc]: Frame::SequenceNum() is an increasing number as time goes. https://github.com/ApolloAuto/apollo/blob/master/modules/planning/common/frame.h#L82

File: WORKSPACE.in [hunk 4]
```diff
@@ -44,6 +44,15 @@ new_local_repository(
     path = "/home/tmp/google_styleguide",
 )
 
+#ad-rss-lib
+new_git_repository(
+    name = "ad_rss_lib",
+    build_file = "third_party/rss_lib.BUILD",
+    tag = "v1.1.0",
```
[the remark this answers is not in the document]
Comment (l1212s) [reply WORKSPACE.in]: @berndgassmann let's stay in v1.1.0 first.

File: third_party/rss_lib.BUILD [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        "src/situation/RSSFormulas.cpp",
+        "src/situation/RssIntersectionChecker.cpp",
+        "src/situation/RSSSituation.cpp",
+        "src/world/RssSituationCoordinateSystemConversion.cpp",
+        "src/world/RssObjectPositionExtractor.cpp",
+    ] + glob(["src/**/*.hpp"]),
+    hdrs = glob(["include/**/*.hpp"]),
+    copts = ["-fPIC","-std=c++11","-Werror","-Wall","-Wextra","-pedantic","-Wconversion", "-Wsign-conversion", "--coverage"],
+    includes = ["include", "include/generated", "src", "tests/test_support"], 
+    linkopts = ["--coverage"],
```
[the remark this answers is not in the document]
Comment (l1212s) [reply third_party/rss_lib.BUILD]: @berndgassmann thanks for your reminder. Apollo team also report issues related with it. I will create a patch to remove redundant link option.

File: modules/planning/tasks/rss/decider_rss.cc [hunk 6]
```diff
@@ -125,94 +132,104 @@ Status RssDecider::Process(Frame *frame,
...
-  rss::core::RssCheck rssCheck;
-  rssCheck.calculateAccelerationRestriction(worldModel,
-                                            accelerationRestriction);
+  Distance dMin_lon =
+      ad_rss::calculateLongitudinalMinSafeDistance(
+      followingObject, leadingObject);
+
+  ad_rss::world::AccelerationRestriction accelerationRestriction;
+  ad_rss::core::RssCheck rssCheck;
+  rssCheck.calculateAccelerationRestriction(
```
[the remark this answers is not in the document]
Comment (l1212s) [reply modules/planning/tasks/rss/decider_rss.cc]: @berndgassmann I respect this comment, will create a patch to fix it.

File: third_party/rss_lib.BUILD [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        "src/physics/Math.cpp",
+        "src/situation/RSSFormulas.cpp",
+        "src/situation/RssIntersectionChecker.cpp",
+        "src/situation/RSSSituation.cpp",
+        "src/world/RssSituationCoordinateSystemConversion.cpp",
+        "src/world/RssObjectPositionExtractor.cpp",
+    ] + glob(["src/**/*.hpp"]),
+    hdrs = glob(["include/**/*.hpp"]),
+    copts = ["-fPIC","-std=c++11","-Werror","-Wall","-Wextra","-pedantic","-Wconversion", "-Wsign-conversion", "--coverage"],
+    includes = ["include", "include/generated", "src", "tests/test_support"],
```
[the remark this answers is not in the document]
Comment (l1212s) [reply third_party/rss_lib.BUILD]: @berndgassmann test_support is still needed for rss_decider_test.cpp

File: modules/planning/tasks/rss/decider_rss.cc [hunk 8]
```diff
@@ -125,94 +132,104 @@ Status RssDecider::Process(Frame *frame,
...
+         << double(laneSegment.length.minimum)
+         << " laneSegment.length.maximum = "
+         << double(laneSegment.length.maximum)
+         << " laneSegment.width.minimum = "
+         << double(laneSegment.width.minimum)
+         << " laneSegment.width.maximum = "
+         << double(laneSegment.width.maximum);
+
+  ad_rss::world::WorldModel worldModel;
+  worldModel.egoVehicle = ad_rss::objectAsEgo(followingObject);
```
[the remark this answers is not in the document]
Comment (l1212s) [reply modules/planning/tasks/rss/decider_rss.cc]: @berndgassmann refer to rss-open-source-library-refine.patch

File: modules/planning/tasks/rss/decider_rss.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (l1212s) [reply modules/planning/tasks/rss/decider_rss.cc]: just submitted the patch "rss-open-source-library-refine" to fix this potential issue.

File: modules/planning/tasks/rss/decider_rss.cc [hunk 9]
```diff
@@ -125,94 +132,104 @@ Status RssDecider::Process(Frame *frame,
...
-  rss::world::AccelerationRestriction accelerationRestriction;
-  rss::core::RssCheck rssCheck;
-  rssCheck.calculateAccelerationRestriction(worldModel,
-                                            accelerationRestriction);
+  Distance dMin_lon =
+      ad_rss::calculateLongitudinalMinSafeDistance(
+      followingObject, leadingObject);
+
+  ad_rss::world::AccelerationRestriction accelerationRestriction;
+  ad_rss::core::RssCheck rssCheck;
```
[the remark this answers is not in the document]
Comment (l1212s) [reply modules/planning/tasks/rss/decider_rss.cc]: @berndgassmann create a patch rss-open-source-library-refine to fix this issue.

File: modules/planning/tasks/rss/decider_rss.h [hunk 10]
```diff
@@ -25,8 +25,8 @@
 #include <string>
 
 #include "modules/planning/tasks/task.h"
-#include "rss/core/RssCheck.hpp"
-#include "rss/test_support/TestSupport.hpp"
+#include "ad_rss/core/RssCheck.hpp"
+#include "TestSupport.hpp"
```
[the remark this answers is not in the document]
Comment (l1212s) [reply modules/planning/tasks/rss/decider_rss.h]: rss_decider needs test_support.h for two reasons:
1. decider_rss_test.cc calls some APIs defined in test_support.h 
2. calculateLongtitudinalMinSafeDistance() of decider_rss.cpp needs test_support helper functions.  This function calculate the safe longitudinal distance shown in Apollo simulator UI.

File: modules/planning/tasks/rss/decider_rss.cc (hunk 1 above)
Comment (l1212s) [reply modules/planning/tasks/rss/decider_rss.cc]: @xiaoxq could you change the review status and the current status is still "request changes".

File: modules/planning/tasks/rss/decider_rss.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (l1212s) [reply modules/planning/tasks/rss/decider_rss.cc]: @berndgassmann submitted a patch "rss-open-library-refine" to fix it.

File: modules/planning/tasks/rss/decider_rss.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (berndgassmann) [reply modules/planning/tasks/rss/decider_rss.h]: One should replace calculateLongtitudinalMinSafeDistance() by the calcluation within RSSFormulas instead. calculateSafeLongitudinalDistanceSameDirection()
For this one has to call the 4 rss functions step by step to access also intermediate results.
On the long run one should perhaps provide the safe distance information in the ResponseState of RSS (-> TODO on ad-rss-lib side).

File: modules/planning/tasks/rss/decider_rss_test.cc [hunk 11]
```diff
@@ -18,51 +18,59 @@
  * @file
  **/
 
-#include "rss/core/RssCheck.hpp"
-#include "rss/test_support/TestSupport.hpp"
 
-namespace rss {
+#include "ad_rss/core/RssCheck.hpp"
+#include "TestSupport.hpp"
```
Comment (berndgassmann) [inline modules/planning/tasks/rss/decider_rss_test.cc]: I still vote strongly to replace TestSupport functionality from production code. use inside the tests is ok for sure. See above on how to handle this.

Comment (l1212s) [reply modules/planning/tasks/rss/decider_rss_test.cc]: @berndgassmann  I removed the test support from decider_rss and please refer to the patch "rss-open-library-wrapper-remove-test-header"
As for decider_rss_test.cc,  let's keep it first and remove the test header in next release.

**Pull request merged**